### ISSUE TITLE
Force sync of table field-values?

### ISSUE THREAD

**Issue by jacktasia** [metabase/metabase#1988]
I have a table that is basically `id`, `date`, `type`, and `count` with `type` being an indexed `string` and `count` being an `int`. When using the GUI question builder and trying to filter by `type` I am only presented with a list of checkboxes for all the distinct values for that column in the table, but this rarely seems to update. This is very problematic if someone enters a new `type` string into this table and does not know how to write SQL since they have no way to build a question around the data they just entered.

Can I manually trigger this list to update? Additionally, it seems like it would be nice if it also allowed an arbitrary filter value for filtering in the GUI since cache validation, in this case, may be easier said than done.

This was with Metabase v0.14.1 with Postgres 9.3

**Comment by agilliland** [comment]
How often are the values in the `type` column changing?  Are you capturing a moving set of data such as the last 48 hours of events such that the data is meant to change with relative frequency?

We currently run a full sync of the database metadata once per day, which isn't exactly frequent, but it avoids overtaxing the db.  You can certainly manually sync that database any time you like by going into the Admin Panel > Databases > Pick Your Db > Click `Sync`.

This is also available via our REST api, so if you wanted to link up a 3rd party trigger to this event you could.  e.g. if an ETL job runs which updates the table you mention you could kick off a quick api call to Metabase and trigger a sync.  That's a lot more involved, but it's the kind of thing you see fairly regularly in a data pipeline.

All this being said, if the data isn't changing all that often then you should be fine with the occasional manual sync of the db.

**Comment by jacktasia** [comment]
> How often are the values in the `type` column changing?

Up to a few time a day.

> Are you capturing a moving set of data such as the last 48 hours of events such that the data is meant to change with relative frequency?

Not exactly. Basically, once a new `type` is added we will want to add two questions with relative dates: 
* Yesterday's single value for that `type`
* "Last 30 days" line graph for the `type`

So ideally, we'd add the first piece of data setup the questions and add to a dashboard and then it will "just work" going forward with new daily data for those `type`s getting added.

> You can certainly manually sync that database any time you like by going into the Admin Panel > Databases > Pick Your Db > Click Sync.

Sorry, I forget to mention that I have tried this a few times, but have yet to see it update. I watched logs and I see the table in question `[******************************************????????] ?   85% Synced table 'stats'.` and it goes up to 100% although I don't see any change in the UI.  

I am curious if this PR https://github.com/metabase/metabase/pull/1981 would also fix my issues? Seems to be touching the parts of the code that I was looking into when I was trying to debug this.

**Comment by agilliland** [comment]
got it.  the PR you mention fixes a bug where the pre-calculated values for a column are not cleared when the data changes to the point where maintaining the values makes sense any more.

if there are more than 40 unique values in your `type` column then that's likely to be what you ran into.

**Comment by camsaul** [comment]
Is this issue resolved?

**Comment by agilliland** [comment]
yeah, i think we can close this one thanks to #1499 which should prevent this from being an issue for folks going forward.

**Comment by jacktasia** [comment]
So I wanted to test your most recent changes to see if I was still running into this issue. I still am, but perhaps not for the same reason. I was watching the logs when I hit the "Sync" button and I've pasted below what I got, but first some context...

1. We have a special metabase postgres role that only has access to tables we want our non-devs to query against. So the credentials metabase gets has this role and they do not have access to some tables 
2. Previously this didn't seem to impact the syncing process, but now it seems to totally kill the syncing process when it first tries to sync a table it can't access. 
3. We have less than the threshold right now (31 unique field values for the column in question).
4. In case this matters: the table in question is called `stats` which is in a DB postgres schema called `stats` (what mysql confusingly calls a "database") this postgres database also has a `stats` table under a different schema. Only potentially important if the table name is used key for a database but the database schema name is being ignored.

Finally the error I see in the logs:
```
02-26 19:45:16 DEBUG driver.sync :: [··················································] 😱    2% Synced table 'annotation_annotation'.
PSQLException:
 Message: ERROR: permission denied for schema grafana
  Position: 33
 SQLState: 42501
 Error Code: 0
02-26 19:45:16 ERROR driver.sync :: Unable to determine row_count for 'api_key': null
PSQLException:
 Message: ERROR: permission denied for schema grafana
  Position: 50
 SQLState: 42501
 Error Code: 0
PSQLException:
 Message: ERROR: permission denied for schema grafana
  Position: 50
 SQLState: 42501
 Error Code: 0
02-26 19:45:16 ERROR events.sync-database :: Error syncing Database: 2 #error {
 :cause Value does not match schema: {:row_count (not (integer? nil))}
 :data {:type :schema.core/error, :schema {#schema.core.OptionalKey{:k :row_count} Int, #schema.core.OptionalKey{:k :fields} [{:id Int, #schema.core.OptionalKey{:k :special-type} (enum :timestamp_seconds :category :zip_code :desc :number :name :city :longitude :state :timestamp_milliseconds :id :avatar :url :latitu
de :image :json :fk :country), #schema.core.OptionalKey{:k :preview-display} java.lang.Boolean, #schema.core.OptionalKey{:k :values} [Any]}]}, :value {:row_count nil, :fields ({:id 802} {:id 797} {:id 800} {:id 799, :values ()} {:id 798} {:id 801, :values ()} {:id 803})}, :error {:row_count (not (integer? nil))}}
 :via
 [{:type clojure.lang.ExceptionInfo
   :message Value does not match schema: {:row_count (not (integer? nil))}
   :data {:type :schema.core/error, :schema {#schema.core.OptionalKey{:k :row_count} Int, #schema.core.OptionalKey{:k :fields} [{:id Int, #schema.core.OptionalKey{:k :special-type} (enum :timestamp_seconds :category :zip_code :desc :number :name :city :longitude :state :timestamp_milliseconds :id :avatar :url :lati
tude :image :json :fk :country), #schema.core.OptionalKey{:k :preview-display} java.lang.Boolean, #schema.core.OptionalKey{:k :values} [Any]}]}, :value {:row_count nil, :fields ({:id 802} {:id 797} {:id 800} {:id 799, :values ()} {:id 798} {:id 801, :values ()} {:id 803})}, :error {:row_count (not (integer? nil))}}
   :at [schema.core$validator$fn__6616 invoke core.clj 151]}]
 :trace
 [[schema.core$validator$fn__6616 invoke core.clj 151]
  [schema.core$validate invokeStatic core.clj 160]
  [schema.core$validate invoke core.clj 155]
  [metabase.driver.sync$sync_database_active_table_BANG_ invokeStatic sync.clj 178]
  [metabase.driver.sync$sync_database_active_table_BANG_ doInvoke sync.clj 158]
  [clojure.lang.RestFn invoke RestFn.java 464]
  [metabase.driver.sync$sync_database_active_tables_BANG_ invokeStatic sync.clj 206]
  [metabase.driver.sync$sync_database_active_tables_BANG_ doInvoke sync.clj 197]
  [clojure.lang.RestFn invoke RestFn.java 464]
  [metabase.driver.sync$sync_database_with_tracking_BANG_ invokeStatic sync.clj 265]
  [metabase.driver.sync$sync_database_with_tracking_BANG_ invoke sync.clj 249]
  [clojure.core$partial$fn__4763 invoke core.clj 2528]
  [metabase.driver$fn__9686 invokeStatic driver.clj 258]
  [metabase.driver$fn__9686 invoke driver.clj 258]
  [metabase.driver$fn__9648$G__9452__9657 invoke driver.clj 82]
  [metabase.driver.sync$sync_database_BANG_ invokeStatic sync.clj 60]
  [metabase.driver.sync$sync_database_BANG_ doInvoke sync.clj 49]
  [clojure.lang.RestFn invoke RestFn.java 464]
  [metabase.driver$sync_database_BANG_ invokeStatic driver.clj 386]
  [metabase.driver$sync_database_BANG_ doInvoke driver.clj 379]
  [clojure.lang.RestFn invoke RestFn.java 410]
  [metabase.events.sync_database$process_sync_database_event$fn__26223 invoke sync_database.clj 32]
  [clojure.core$binding_conveyor_fn$fn__4676 invoke core.clj 1938]
  [clojure.lang.AFn call AFn.java 18]
  [java.util.concurrent.FutureTask run FutureTask.java 262]
  [java.util.concurrent.ThreadPoolExecutor runWorker ThreadPoolExecutor.java 1145]
  [java.util.concurrent.ThreadPoolExecutor$Worker run ThreadPoolExecutor.java 615]
  [java.lang.Thread run Thread.java 745]]}
```

Please let me know if I can expand on anything. Thanks!

**Comment by agilliland** [comment]
The error you are seeing has actually been fixed already.  Earlier this week we created a branch for `release-0.15.0` which contains the code in preparation for release in a few days.  There are bug fixes in that branch (including this one) which aren't in master right now.

So I suspect that if you checkout `release-0.15.0` and build from that you'll have better luck.

**Comment by jacktasia** [comment]
> There are bug fixes in that branch (including this one) which aren't in master right now.

Oh, ok. Cool. Thanks! I will try that now and let you know.

**Comment by jacktasia** [comment]
> So I suspect that if you checkout release-0.15.0 and build from that you'll have better luck.

Yup. It's working as expected now. :+1:  Thanks!

**Comment by camsaul** [comment]
Closing since resolved in `0.15.0`
